### PR TITLE
feat: implement cli

### PR DESCRIPTION
**Pull request by pionxzh**
#### NPM
image

#### PNPM

image

#### Yarn

image

#### Bun

image

---

Tested with the following combinations:
- [X] NPM V8
- [ ] NPM Workspace
- [X] PNPM
- [X] PNPM Workspace
- [x] Yarn V1
- [ ] Yarn V2
- [X] Yarn V3
- [ ] Yarn Workspace

### CODE REVIEW

File: packages/cli/src/index.ts [hunk 1]
```diff
@@ -34,9 +46,31 @@ process.on('SIGTERM', handleSigTerm);
...
+        const packageJsonPath = path.join(projectPath, 'package.json');
+        const packageJson = await fs.readFile(packageJsonPath, 'utf-8');
+        const parsedPackageJson = JSON.parse(packageJson);
+        parsedPackageJson.overrides = {
+          ...parsedPackageJson.overrides,
+          ...overrides
+        };
+        // TODO: respect existing format?
+        // TODO: confirm with user
+        await fs.writeFile(packageJsonPath, JSON.stringify(parsedPackageJson, null, 2));
```
Comment (SukkaW) [inline packages/cli/src/index.ts:73]: pnpm uses `.pnpm.overrides`, not `.overrides`: https://pnpm.io/package_json#pnpmoverrides

yarn uses `.resolutions`: https://yarnpkg.com/configuration/manifest#resolutions

Need to handle them separately.

Comment (pionxzh) [reply packages/cli/src/index.ts]: updated

**Comment by pionxzh** [comment]
`Bun` support is blocked by https://github.com/oven-sh/bun/issues/1134

File: packages/cli/src/index.ts [hunk 2]
```diff
@@ -1,42 +1,159 @@
...
+      delete node.dependencies;
+      packagesToBeOverride.push(node);
+    } else if (node.dependencies) {
+      node.dependencies.forEach(child => traverse(child));
+    }
+  };
+
+  searchResult.forEach(node => traverse(node));
+
+  packagesToBeOverride.splice(0, 18);
```
[the remark this answers is not in the document]
Comment (pionxzh) [reply packages/cli/src/index.ts:63]: I use it to reduce the length for screenshot 🤣 my bad

File: packages/cli/src/index.ts [hunk 3]
```diff
@@ -1,42 +1,159 @@
...
-          .default('auto', 'detect package manager automatically')
-      )
-      .description('replace redundant polyfills inside your project\'s node_modules with nolyfill')
-      .action((source: string | undefined, option: unknown) => {
-        console.log(source, option);
-        throw new Error('not implemented');
+      .addOption(pmCommandOption)
+      .action(async (source: string | undefined, option: PmCommandOptions) => {
+        const projectPath = path.resolve(source ?? process.cwd());
+        const packageManager = option.pm === 'auto' ? await detectPackageManager(projectPath) : option.pm;
```
[the remark this answers is not in the document]
Comment (pionxzh) [reply packages/cli/src/index.ts:134]: What would be the expected behavior of `bun` for you?
1. should we still read and check bun lockfile?
2. should we still write `overrides` to package.json?

[the remark this answers is not in the document]
Comment (pionxzh) [reply packages/cli/src/index.ts:133]: image

**Review by SukkaW — APPROVED** [review]

**Comment by SukkaW** [comment]
telegram-cloud-photo-size-5-6278375170750134623-x

**Pull request merged**